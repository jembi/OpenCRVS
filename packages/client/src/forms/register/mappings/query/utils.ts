/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { IFormData } from '@client/forms'
import { GQLRegWorkflow } from '@opencrvs/gateway/src/graphql/schema'

export function transformStatusData(
  transformedData: IFormData,
  statusData: GQLRegWorkflow[],
  sectionId: string
) {
  const registrationStatus =
    statusData &&
    statusData.find((status) => {
      return status.type && status.type === 'REGISTERED'
    })
  transformedData[sectionId] = {
    ...transformedData[sectionId],
    commentsOrNotes:
      (statusData &&
        statusData[0] &&
        statusData[0].comments &&
        statusData[0].comments[0] &&
        statusData[0].comments[0].comment) ||
      ''
  }

  if (!registrationStatus) {
    return transformedData
  }
  transformedData[sectionId] = {
    ...transformedData[sectionId],
    regStatus: {
      type: registrationStatus.type || '',
      statusDate: registrationStatus.timestamp,
      officeName:
        (registrationStatus.office && registrationStatus.office.name) || '',
      officeAlias:
        (registrationStatus.office &&
          registrationStatus.office.alias &&
          registrationStatus.office.alias.join(' ')) ||
        '',
      officeAddressLevel3:
        (registrationStatus.office &&
          registrationStatus.office.address &&
          registrationStatus.office.address.district) ||
        '',
      officeAddressLevel4:
        (registrationStatus.office &&
          registrationStatus.office.address &&
          registrationStatus.office.address.state) ||
        ''
    }
  }
  return transformedData
}
