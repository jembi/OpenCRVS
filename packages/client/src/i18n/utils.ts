/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors. OpenCRVS and the OpenCRVS
 * graphic logo are (registered/a) trademark(s) of Plan International.
 */
import * as queryString from 'querystring'
import { storage } from '@client/storage'

export function getAvailableLanguages() {
  return window.config.LANGUAGES.split(',')
}

export function getDefaultLanguage() {
  return getAvailableLanguages()[0]
}

export async function getPreferredLanguage() {
  const languageInUrl = queryString.parse(
    window.location.search.replace(/^\?/, '')
  ).language as string | undefined

  const alreadyStoredLanguage = await storage.getItem('language')

  if (languageInUrl) {
    return languageInUrl
  } else if (alreadyStoredLanguage) {
    return alreadyStoredLanguage
  } else {
    return getDefaultLanguage()
  }
}

export function storeLanguage(language: string) {
  storage.setItem('language', language)
}
