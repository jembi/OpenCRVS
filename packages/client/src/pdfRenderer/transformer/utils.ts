/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors. OpenCRVS and the OpenCRVS
 * graphic logo are (registered/a) trademark(s) of Plan International.
 */
import {
  ExecutorKey,
  IEventWiseKey,
  Condition,
  ConditionOperation
} from '@client/pdfRenderer/transformer/types'
import {
  Event,
  IFormData,
  IFormSectionData,
  IFormFieldValue
} from '@opencrvs/client/src/forms'
import { IApplication } from '@client/applications'
import { MessageDescriptor } from 'react-intl'

const eventMessageDescriptor = {
  [Event.BIRTH]: {
    defaultMessage: 'Birth',
    description: 'A label from the birth event',
    id: 'constants.birth'
  },
  [Event.DEATH]: {
    defaultMessage: 'Death',
    description: 'A label from the death event',
    id: 'constants.death'
  }
}
export function getValueFromApplicationDataByKey(
  data: IFormData,
  valueKey: string
) {
  const keyTree: string[] = valueKey.split('.')

  let valueObject: IFormSectionData | IFormFieldValue | null = null

  keyTree.forEach((keyNode) => {
    valueObject =
      valueObject === null
        ? data[keyNode] || null
        : (valueObject as IFormSectionData)[keyNode]
  })
  if (valueObject === null) {
    // eslint-disable-next-line no-console
    console.error(`Given value key structure is not valid: ${valueKey}`)
  }
  return valueObject
}

export function getEventMessageDescription(event: Event): MessageDescriptor {
  return eventMessageDescriptor[event]
}

export function getExecutorKeyValue(key: ExecutorKey) {
  if (key === 'CURRENT_DATE') {
    return Date.now()
  }
  throw new Error('Invalid executor key found')
}

export function getExecutorFieldValue(
  key: ExecutorKey | IEventWiseKey,
  application: IApplication
) {
  let value = null
  if (typeof key === 'string') {
    value = getExecutorKeyValue(key as ExecutorKey)
  } else {
    value = getValueFromApplicationDataByKey(
      application.data,
      (key as IEventWiseKey)[application.event]
    )
  }
  return value
}

function executeConditionalOperation(
  operation: ConditionOperation,
  conditionInputs: string[],
  value: string
) {
  switch (operation) {
    case ConditionOperation.VALUE_EXISTS:
      return value !== ''
    case ConditionOperation.VALUE_DOES_NOT_EXISTS:
      return value === ''
    case ConditionOperation.DOES_NOT_MATCH:
      return !conditionInputs.includes(value)
    default:
      return conditionInputs.includes(value)
  }
}

function getConditionalOperationEnumByValue(value: string | undefined) {
  if (!value) {
    return ConditionOperation.MATCH
  }
  return ConditionOperation[value as keyof typeof ConditionOperation]
}

export function getMatchedCondition(
  conditions: Condition[],
  applicationData: IFormData
) {
  return conditions.find((conditionObj) => {
    try {
      if (!conditionObj.condition) {
        return true
      }
      return executeConditionalOperation(
        getConditionalOperationEnumByValue(conditionObj.condition.operation),
        conditionObj.condition.values,
        // Will return empty string when value is not found for given key
        getValueFromApplicationDataByKey(
          applicationData,
          conditionObj.condition.key || ''
        ) || ''
      )
    } catch (error) {
      /* eslint-disable no-console */
      console.info(`PDF UTILS ERROR: ${error}`)
      /* eslint-enable no-console */
      return false
    }
  })
}
