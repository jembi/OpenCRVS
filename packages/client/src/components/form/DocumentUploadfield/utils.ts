import { getBase64String, ErrorMessage } from './DocumentUploaderWithOption'
import Jimp from 'jimp'
import { buttonMessages, formMessages as messages } from '@client/i18n/messages'
import { IFormFieldValue, IAttachmentValue } from '@client/forms'

export const handleFileChange = (thisObj: any, uploadedImage: File) => {
  if (!uploadedImage) {
    return
  }
  const allowedDocType = thisObj.props.allowedDocType

  thisObj.setState(() => ({
    filesBeingUploaded: [
      ...thisObj.state.filesBeingUploaded,
      {
        label: uploadedImage.name
      }
    ]
  }))

  thisObj.props.onUploadingStateChanged &&
    thisObj.props.onUploadingStateChanged(true)

  getBase64String(uploadedImage).then((data) => {
    let base64String = data as string
    base64String = base64String.split('base64,')[1]
    Jimp.read(new Buffer(base64String, 'base64'))
      .then((buffer) => {
        if (
          allowedDocType &&
          allowedDocType.length > 0 &&
          !allowedDocType.includes(buffer.getMIME())
        ) {
          throw new Error('File type not supported')
        }
        return data as string
      })
      .then((buffer) => {
        thisObj.props.onUploadingStateChanged &&
          thisObj.props.onUploadingStateChanged(false)
        thisObj.props.onComplete({
          name: uploadedImage.name,
          type: uploadedImage.type,
          data: buffer
        })
        thisObj.setState({
          error: ''
        })
        thisObj.setState({
          filesBeingUploaded: []
        })
      })
      .catch(() => {
        thisObj.props.onUploadingStateChanged &&
          thisObj.props.onUploadingStateChanged(false)
        thisObj.setState({
          filesBeingUploaded: []
        })
        allowedDocType &&
          allowedDocType.length > 0 &&
          thisObj.setState({
            error: thisObj.props.intl.formatMessage(messages.fileUploadError, {
              type: allowedDocType
                .map((docTypeStr: string) => docTypeStr.split('/').pop())
                .join(', ')
            })
          })
      })
  })
}

export const selectForPreview = (
  thisObj: any,
  previewImage: IFormFieldValue
) => {
  thisObj.setState({ previewImage: previewImage as IAttachmentValue })
}

export const closePreviewSection = (thisObj: any) => {
  thisObj.setState({ previewImage: null })
}

export const onDelete = (thisObj: any, image: IFormFieldValue) => {
  thisObj.props.onComplete('')
  thisObj.closePreviewSection()
}
