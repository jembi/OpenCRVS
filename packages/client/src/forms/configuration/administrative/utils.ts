/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors. OpenCRVS and the OpenCRVS
 * graphic logo are (registered/a) trademark(s) of Plan International.
 */

import { sentenceCase } from '@client/utils/data-formatting'
import {
  AddressCases,
  EventLocationAddressCases
} from '@client/forms/configuration/administrative/addresses'
import { IConditional, SerializedFormField } from '@client/forms/index'
import { Address } from '@client/utils/gateway'

export function getDependency(location: string, useCase: string) {
  switch (location) {
    case 'state':
      return useCase === 'placeOfBirth' ||
        useCase === 'placeOfDeath' ||
        useCase === 'placeOfMarriage'
        ? 'country'
        : `country${sentenceCase(useCase)}`
    case 'district':
      return useCase === 'placeOfBirth' ||
        useCase === 'placeOfDeath' ||
        useCase === 'placeOfMarriage'
        ? 'state'
        : `state${sentenceCase(useCase)}`
    case 'locationLevel3':
      return useCase === 'placeOfBirth' ||
        useCase === 'placeOfDeath' ||
        useCase === 'placeOfMarriage'
        ? 'district'
        : `district${sentenceCase(useCase)}`
    case 'locationLevel4':
      return useCase === 'placeOfBirth' ||
        useCase === 'placeOfDeath' ||
        useCase === 'placeOfMarriage'
        ? 'locationLevel3'
        : `locationLevel3${sentenceCase(useCase)}`
    case 'locationLevel5':
      return useCase === 'placeOfBirth' ||
        useCase === 'placeOfDeath' ||
        useCase === 'placeOfMarriage'
        ? 'locationLevel4'
        : `locationLevel4${sentenceCase(useCase)}`
  }
}

export function getRuralOrUrbanConditionals(
  useCase: string,
  defaultConditionals: IConditional[]
) {
  let customConditionals: IConditional[] = []
  switch (window.config.ADMIN_LEVELS) {
    case 1:
      customConditionals = [
        {
          action: 'hide',
          expression: `!values.state${sentenceCase(useCase)}`
        }
      ]
      break
    case 2:
      customConditionals = [
        {
          action: 'hide',
          expression: `!values.state${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.district${sentenceCase(useCase)}`
        }
      ]
      break
    case 3:
      customConditionals = [
        {
          action: 'hide',
          expression: `!values.state${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.district${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.locationLevel3${sentenceCase(useCase)}`
        }
      ]
      break
    case 4:
      customConditionals = [
        {
          action: 'hide',
          expression: `!values.state${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.district${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.locationLevel3${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.locationLevel4${sentenceCase(useCase)}`
        }
      ]
      break
  }
  return defaultConditionals.concat(customConditionals)
}

export function getConditionals(location: string, useCase: string) {
  switch (location) {
    case 'state':
      return [
        {
          action: 'hide',
          expression: `!values.country${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!isDefaultCountry(values.country${sentenceCase(
            useCase
          )})`
        }
      ]
    case 'district':
      return [
        {
          action: 'hide',
          expression: `!values.country${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.state${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!isDefaultCountry(values.country${sentenceCase(
            useCase
          )})`
        }
      ]
    case 'locationLevel3':
      return [
        {
          action: 'hide',
          expression: `!values.country${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.state${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.district${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!isDefaultCountry(values.country${sentenceCase(
            useCase
          )})`
        }
      ]
    case 'locationLevel4':
      return [
        {
          action: 'hide',
          expression: `!values.country${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.state${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.district${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.locationLevel3${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!isDefaultCountry(values.country${sentenceCase(
            useCase
          )})`
        }
      ]
    case 'locationLevel5':
      return [
        {
          action: 'hide',
          expression: `!values.country${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.state${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.district${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.locationLevel3${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!values.locationLevel4${sentenceCase(useCase)}`
        },
        {
          action: 'hide',
          expression: `!isDefaultCountry(values.country${sentenceCase(
            useCase
          )})`
        }
      ]
  }
}

export function getLocationSelect(
  location: string,
  useCase: string,
  locationIndex: number,
  informant: boolean
): SerializedFormField {
  return {
    name: `${location}${sentenceCase(useCase)}`,
    type: 'SELECT_WITH_DYNAMIC_OPTIONS',
    label: {
      defaultMessage: '',
      description: `Title for the ${location} select`,
      id: `form.field.label.${location}`
    },
    customisable: false,
    previewGroup: `${useCase}Address`,
    required: true,
    initialValue: '',
    validate: [],
    placeholder: {
      defaultMessage: 'Select',
      description: 'Placeholder text for a select',
      id: 'form.field.select.placeholder'
    },
    dynamicOptions: {
      resource: 'locations',
      dependency: getDependency(location, useCase),
      initialValue: 'agentDefault'
    },
    conditionals: getConditionals(location, useCase),
    mapping: {
      mutation: informant
        ? {
            operation: 'fieldValueNestingTransformer',
            parameters: [
              'individual',
              {
                operation: 'fieldToAddressTransformer',
                parameters: [
                  useCase.toUpperCase() === 'PRIMARY'
                    ? AddressCases.PRIMARY_ADDRESS
                    : AddressCases.SECONDARY_ADDRESS,
                  locationIndex,
                  location
                ]
              },
              'address'
            ]
          }
        : {
            operation: 'fieldToAddressTransformer',
            parameters: [
              useCase.toUpperCase() === 'PRIMARY'
                ? AddressCases.PRIMARY_ADDRESS
                : AddressCases.SECONDARY_ADDRESS,
              locationIndex,
              location
            ]
          },
      query: informant
        ? {
            operation: 'nestedValueToFieldTransformer',
            parameters: [
              'individual',
              {
                operation: 'addressToFieldTransformer',
                parameters: [
                  useCase.toUpperCase() === 'PRIMARY'
                    ? AddressCases.PRIMARY_ADDRESS
                    : AddressCases.SECONDARY_ADDRESS,
                  locationIndex,
                  location
                ]
              }
            ]
          }
        : {
            operation: 'addressToFieldTransformer',
            parameters: [
              useCase.toUpperCase() === 'PRIMARY'
                ? AddressCases.PRIMARY_ADDRESS
                : AddressCases.SECONDARY_ADDRESS,
              locationIndex,
              location
            ]
          }
    }
  }
}

export function getPlaceOfEventConditionals(
  location: string,
  configCase: EventLocationAddressCases
) {
  switch (location) {
    case 'state':
      return [
        {
          action: 'hide',
          expression: '!values.country'
        },
        {
          action: 'hide',
          expression:
            configCase !== EventLocationAddressCases.PLACE_OF_MARRIAGE
              ? `(values.${configCase}!="OTHER" && values.${configCase}!="PRIVATE_HOME")`
              : ''
        },
        {
          action: 'hide',
          expression: '!isDefaultCountry(values.country)'
        }
      ]
    case 'district':
      return [
        {
          action: 'hide',
          expression: '!values.country'
        },
        {
          action: 'hide',
          expression: '!values.state'
        },
        {
          action: 'hide',
          expression:
            configCase !== EventLocationAddressCases.PLACE_OF_MARRIAGE
              ? `(values.${configCase}!="OTHER" && values.${configCase}!="PRIVATE_HOME")`
              : ''
        },
        {
          action: 'hide',
          expression: '!isDefaultCountry(values.country)'
        }
      ]
    case 'locationLevel3':
      return [
        {
          action: 'hide',
          expression: '!values.country'
        },
        {
          action: 'hide',
          expression: '!values.state'
        },
        {
          action: 'hide',
          expression: '!values.district'
        },
        {
          action: 'hide',
          expression:
            configCase !== EventLocationAddressCases.PLACE_OF_MARRIAGE
              ? `(values.${configCase}!="OTHER" && values.${configCase}!="PRIVATE_HOME")`
              : ''
        },
        {
          action: 'hide',
          expression: '!isDefaultCountry(values.country)'
        }
      ]
    case 'locationLevel4':
      return [
        {
          action: 'hide',
          expression: '!values.country'
        },
        {
          action: 'hide',
          expression: '!values.state'
        },
        {
          action: 'hide',
          expression: '!values.district'
        },
        {
          action: 'hide',
          expression: '!values.locationLevel3'
        },
        {
          action: 'hide',
          expression:
            configCase !== EventLocationAddressCases.PLACE_OF_MARRIAGE
              ? `(values.${configCase}!="OTHER" && values.${configCase}!="PRIVATE_HOME")`
              : ''
        },
        {
          action: 'hide',
          expression: '!isDefaultCountry(values.country)'
        }
      ]
    case 'locationLevel5':
      return [
        {
          action: 'hide',
          expression: '!values.country'
        },
        {
          action: 'hide',
          expression: '!values.state'
        },
        {
          action: 'hide',
          expression: '!values.district'
        },
        {
          action: 'hide',
          expression: '!values.locationLevel3'
        },
        {
          action: 'hide',
          expression: '!values.locationLevel4'
        },
        {
          action: 'hide',
          expression:
            configCase !== EventLocationAddressCases.PLACE_OF_MARRIAGE
              ? `(values.${configCase}!="OTHER" && values.${configCase}!="PRIVATE_HOME")`
              : ''
        },
        {
          action: 'hide',
          expression: '!isDefaultCountry(values.country)'
        }
      ]
  }
}

export function getPlaceOfEventLocationSelect(
  location: string,
  configCase: EventLocationAddressCases,
  locationIndex: number
): SerializedFormField {
  return {
    name: location,
    customisable: false,
    type: 'SELECT_WITH_DYNAMIC_OPTIONS',
    label: {
      defaultMessage: sentenceCase(location),
      description: `Title for the ${location} select`,
      id: `form.field.label.${location}`
    },
    previewGroup: configCase,
    required: true,
    initialValue: '',
    validate: [],
    placeholder: {
      defaultMessage: 'Select',
      description: 'Placeholder text for a select',
      id: 'form.field.select.placeholder'
    },
    dynamicOptions: {
      resource: 'locations',
      dependency: getDependency(location, configCase),
      initialValue: 'agentDefault'
    },
    conditionals: getPlaceOfEventConditionals(location, configCase),
    mapping: {
      template: {
        fieldName: configCase,
        operation: 'eventLocationAddressOfflineTransformer',
        parameters: [location, configCase]
      },
      mutation: {
        operation:
          configCase === EventLocationAddressCases.PLACE_OF_BIRTH
            ? 'birthEventLocationMutationTransformer'
            : configCase === EventLocationAddressCases.PLACE_OF_DEATH
            ? 'deathEventLocationMutationTransformer'
            : 'marriageEventLocationMutationTransformer',
        parameters: [locationIndex]
      },
      query: {
        operation: 'eventLocationQueryTransformer',
        parameters: [
          locationIndex,
          location,
          {
            fieldsToIgnoreForLocalAddress: [
              'internationalDistrict',
              'internationalState'
            ],
            fieldsToIgnoreForInternationalAddress: [
              'locationLevel3',
              'locationLevel4',
              'locationLevel5',
              'district',
              'state'
            ]
          }
        ]
      }
    }
  }
}
