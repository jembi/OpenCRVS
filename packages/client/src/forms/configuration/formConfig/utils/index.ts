/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors. OpenCRVS and the OpenCRVS
 * graphic logo are (registered/a) trademark(s) of Plan International.
 */

import {
  IFormField,
  IRadioOption,
  ISelectOption,
  RADIO_GROUP_WITH_NESTED_FIELDS,
  RADIO_GROUP,
  SELECT_WITH_OPTIONS,
  DOCUMENT_UPLOADER_WITH_OPTION,
  ISerializedForm,
  IRadioGroupFormField,
  IRadioGroupWithNestedFieldsFormField,
  ISelectFormFieldWithOptions,
  IDocumentUploaderWithOptionsFormField,
  SerializedFormField,
  IFormDataSet
} from '@client/forms'
import {
  IQuestionConfig,
  isDefaultQuestionConfig,
  getConfiguredQuestions,
  getIdentifiersFromFieldId,
  ICustomSelectOption,
  ICustomQuestionConfig
} from '@client/forms/questionConfig'
import { CustomFieldType, Event, QuestionInput } from '@client/utils/gateway'
import { FieldPosition } from '@client/forms/configuration'
import { deserializeFormField } from '@client/forms/mappings/deserializer'
import { createCustomField } from '@client/forms/configuration/customUtils'
import {
  isPreviewGroupConfigField,
  IPreviewGroupConfigField
} from './previewGroup'
import { ICustomConfigField, isCustomConfigField } from './customConfig'
import {
  IDefaultConfigField,
  defaultQuestionToConfigField,
  isDefaultConfigField,
  hasDefaultFieldChanged
} from './defaultConfig'
import { getField } from '@client/forms/configuration/defaultUtils'
import { ISelectOption as IDataSourceOption } from '@opencrvs/components/lib/Select'

export * from './previewGroup'
export * from './customConfig'
export * from './defaultConfig'

export type IConfigField =
  | IDefaultConfigField
  | ICustomConfigField
  | IPreviewGroupConfigField

export type ISectionFieldMap = Record<string, IConfigField[]>

export type IDataSourceSelectOption = IDataSourceOption & {
  options: ICustomSelectOption[] | ISelectOption[]
}

export function getFieldDefinition(
  configField: IDefaultConfigField | ICustomConfigField,
  defaultForm: ISerializedForm
) {
  let serializedField: SerializedFormField
  if (isDefaultConfigField(configField)) {
    serializedField = {
      ...getField(configField.identifiers, defaultForm),
      required: configField.required
    }
  } else {
    serializedField = createCustomField(configField)
  }
  /* We need to build the field regardless of the conditionals */
  delete serializedField.conditionals
  return deserializeFormField(serializedField)
}

function hasOptions(
  formField: IFormField
): formField is
  | IRadioGroupFormField
  | IRadioGroupWithNestedFieldsFormField
  | ISelectFormFieldWithOptions
  | IDocumentUploaderWithOptionsFormField {
  if (
    formField.type === RADIO_GROUP ||
    formField.type === RADIO_GROUP_WITH_NESTED_FIELDS ||
    formField.type === SELECT_WITH_OPTIONS ||
    formField.type === DOCUMENT_UPLOADER_WITH_OPTION
  ) {
    return true
  }
  return false
}

function getContentKey(formField: IFormField) {
  let contentKeys = [formField.label.id]
  if (
    hasOptions(formField) &&
    !['nationality', 'country', 'countryPrimary', 'countrySecondary'].includes(
      formField.name
    )
  ) {
    contentKeys = contentKeys.concat(
      /* We can remove this type assertion in typescript 4.2+ */
      (formField.options as (ISelectOption | IRadioOption)[]).map(
        (option) => option.label.id
      )
    )
  }
  return contentKeys
}

export function getContentKeys(
  configField: IDefaultConfigField | IPreviewGroupConfigField,
  defaultForm: ISerializedForm
) {
  return isDefaultConfigField(configField)
    ? getContentKey(getFieldDefinition(configField, defaultForm))
    : configField.configFields
        .map((field) => getContentKey(getFieldDefinition(field, defaultForm)))
        .flat()
}

export function getCertificateHandlebar(formField: IFormField) {
  return formField.mapping?.template?.[0]
}

function resolveOptionsFormDatasetId(
  question: ICustomQuestionConfig,
  formDataset: IFormDataSet[]
) {
  if (question.fieldType === CustomFieldType.SelectWithOptions) {
    const options =
      formDataset.find((dataset) => dataset._id === question.datasetId)
        ?.options || []
    return {
      ...question,
      options
    }
  }
  return question
}

export function generateConfigFields(
  event: Event,
  defaultForm: ISerializedForm,
  questions: IQuestionConfig[],
  formDataset: IFormDataSet[]
) {
  questions = questions.filter((question) => question.fieldId.startsWith(event))

  const fields = getConfiguredQuestions(event, defaultForm, questions)
    .map((sectionQuestionConfigs) =>
      sectionQuestionConfigs
        .reduce<IConfigField[]>(
          (configFields, question) => [
            ...configFields,
            isDefaultQuestionConfig(question)
              ? defaultQuestionToConfigField(question, defaultForm)
              : resolveOptionsFormDatasetId(question, formDataset)
          ],
          []
        )
        .reduce<IConfigField[]>((configFields, currentConfigField, index) => {
          if (!index) return [currentConfigField]
          const previousConfigField = configFields[configFields.length - 1]
          if (
            /**
             *
             * Bypass handle preview group for
             * question config in case of custom question
             *
             */
            !isCustomConfigField(currentConfigField) &&
            isPreviewGroupConfigField(previousConfigField) &&
            isPreviewGroupConfigField(currentConfigField) &&
            previousConfigField.previewGroup === currentConfigField.previewGroup
          ) {
            previousConfigField.configFields = [
              ...previousConfigField.configFields,
              ...currentConfigField.configFields
            ]
            return configFields
          }
          return [...configFields, currentConfigField]
        }, [])
    )
    .reduce<ISectionFieldMap>((sectionMap, sectionConfigFields) => {
      if (sectionConfigFields.length === 0) return sectionMap
      const { sectionId } = getIdentifiersFromFieldId(
        sectionConfigFields[0].fieldId
      )
      return {
        ...sectionMap,
        [sectionId]: sectionConfigFields
      }
    }, {})

  return fields
}

function configFieldsToQuestionConfigs(
  configFieldsMap: ISectionFieldMap
): IQuestionConfig[] {
  return Object.values(configFieldsMap).flatMap((sectionConfigFields) =>
    sectionConfigFields
      .flatMap((configField) => {
        if (
          isDefaultConfigField(configField) ||
          isCustomConfigField(configField)
        )
          return [configField]
        return configField.configFields
      })
      .map((configField, index, inputConfigFields) => ({
        ...configField,
        precedingFieldId: !index
          ? FieldPosition.TOP
          : inputConfigFields[index - 1].fieldId
      }))
  )
}

export function generateModifiedQuestionConfigs(
  configFields: ISectionFieldMap,
  defaultRegisterForm: ISerializedForm
): QuestionInput[] {
  const questionConfigs = configFieldsToQuestionConfigs(configFields)

  const fields = questionConfigs
    .filter((questionConfig) => {
      if (isDefaultQuestionConfig(questionConfig)) {
        return hasDefaultFieldChanged(questionConfig, defaultRegisterForm)
      }
      return true
    })
    .map((questionConfig) => {
      if (isDefaultQuestionConfig(questionConfig)) {
        const { identifiers, ...rest } = questionConfig
        return rest
      }

      delete questionConfig.options
      return questionConfig as QuestionInput
    })

  return fields as QuestionInput[]
}
