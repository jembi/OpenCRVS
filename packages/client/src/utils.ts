/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { storage } from '@client/storage'
import { APPLICATION_VERSION, LANG_EN } from '@client/utils/constants'
import { IUserData } from '@client/declarations'
import React from 'react'
import { GetSystemRolesQuery, Role } from '@client/utils/gateway'

export async function validateApplicationVersion() {
  const runningVer = localStorage.getItem('running-version')

  if (!runningVer || runningVer !== APPLICATION_VERSION) {
    localStorage.setItem('running-version', APPLICATION_VERSION)
    const userData = await storage.getItem('USER_DATA')
    const allUserData: IUserData[] = !userData
      ? []
      : (JSON.parse(userData) as IUserData[])

    allUserData.forEach((userData) => {
      userData['declarations'] = []
    })

    await storage.setItem('USER_DATA', JSON.stringify(allUserData))
  }
}
export function isNavigatorOnline() {
  return navigator.onLine
}

export function useOnlineStatus() {
  const [isOnline, setOnline] = React.useState(isNavigatorOnline())
  const ONLINE_CHECK_INTERVAL = 500
  React.useEffect(() => {
    const intervalID = setInterval(
      () => setOnline(isNavigatorOnline()),
      ONLINE_CHECK_INTERVAL
    )

    return () => clearInterval(intervalID)
  }, [])
  return isOnline
}

export function getUserRole(lang: string, role: Role) {
  const defaultLabel = role?.labels?.find((label) => label.lang === LANG_EN)
  const label = role?.labels?.find((label) => label.lang === lang)
  return label?.label || defaultLabel?.label
}

export type RolesInput = (Omit<Role, '_id'> & { _id?: string })[]

export type ISystemRole = NonNullable<
  GetSystemRolesQuery['getSystemRoles']
>[number]
