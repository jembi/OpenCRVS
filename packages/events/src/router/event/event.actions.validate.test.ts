/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { createTestClient, setupTestCase } from '@events/tests/utils'
import { ActionType, SCOPES } from '@opencrvs/commons'

test('Validation error message contains all the offending fields', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user, [SCOPES.RECORD_REGISTER])

  const event = await client.event.create(generator.event.create())

  const data = generator.event.actions.register(event.id, {
    data: {
      'applicant.dob': '02-02'
    }
  })

  await expect(client.event.actions.register(data)).rejects.matchSnapshot()
})

test('Action without form definition should accept empty payload', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  const data = generator.event.actions.validate(event.id, {
    data: {}
  })

  const result = await client.event.actions.validate(data)
  const lastAction = result.actions[result.actions.length - 1]

  expect(lastAction.type).toBe(ActionType.VALIDATE)
  expect(lastAction.data).toEqual({})
})
