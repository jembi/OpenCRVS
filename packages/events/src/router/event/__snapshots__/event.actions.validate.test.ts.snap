// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Validation error message contains all the offending fields 1`] = `[TRPCError: [{"message":"Required for registration","id":"applicant.firstname"},{"message":"Required for registration","id":"applicant.surname"},{"message":"Please enter a valid date","id":"applicant.dob","value":"02-02"},{"message":"Invalid date. Please use the format YYYY-MM-DD","id":"applicant.dob","value":"02-02"}]]`;

exports[`prevents forbidden access if missing required scope 1`] = `[TRPCError: FORBIDDEN]`;
